### PR TITLE
Fix #76 Add Oauth for Search API

### LINKED ISSUE #76: Artist Not Found
First noticed issue when a command that previously worked stopped working.
"spotify play artist moderat" now returns "no results when searching for moderat"
Issue persists for all searches. "spotify play artist drake" also returns that no results were found. 

```
Searching artists for: drake
No results when searching for drake

### PR DESCRIPTION
**Pull request by fonsecapeter**
Just hit the token endpoint before hitting the search endpoint

### CODE REVIEW

File: spotify [hunk 1]
```diff
@@ -119,18 +120,45 @@ while [ $# -gt 0 ]; do
                 array=( $@ );
                 len=${#array[@]};
                 SPOTIFY_SEARCH_API="https://api.spotify.com/v1/search"
+                readonly SPOTIFY_TOKEN_URI="https://accounts.spotify.com/api/token"
+                readonly SHPOTIFY_CLIENT_ID="a939e7a789b1444fad21b8e14b62628a"
```
Comment (hnarayanan) [inline spotify]: Is it ok that you've embedded these in the source code?

File: spotify [hunk 2]
```diff
@@ -119,18 +120,45 @@ while [ $# -gt 0 ]; do
...
+                readonly SHPOTIFY_CLIENT_SECRET="my_client_secret"
```
Comment (hnarayanan) [inline spotify]: Does this value not matter?

File: spotify [hunk 3]
```diff
@@ -119,18 +120,45 @@ while [ $# -gt 0 ]; do
...
+                readonly SHPOTIFY_CREDENTIALS=$(printf "${SHPOTIFY_CLIENT_ID}:${SHPOTIFY_CLIENT_SECRET}" | base64 --wrap=0)
                 SPOTIFY_PLAY_URI="";
 
                 searchAndPlay() {
                     type="$1"
                     Q="$2"
 
+                    cecho "Authorizing API"
+                    cecho "  getting token..."
```
Comment (hnarayanan) [inline spotify]: Please make sure this looks like the rest of the output when you use the app.

File: spotify [hunk 4]
```diff
@@ -119,18 +120,45 @@ while [ $# -gt 0 ]; do
...
+                            -H "Authorization: Basic ${SHPOTIFY_CREDENTIALS}" \
+                            -d "grant_type=client_credentials" \
+                    )
+                    if ! [[ "${SPOTIFY_TOKEN_RESPONSE_DATA}" =~ "access_token" ]]; then
+                        cecho "  Woops, we're having problems connecting to the spotify API."
+                        cecho "  => ${SPOTIFY_TOKEN_RESPONSE_DATA}"
+                        exit 1
+                    fi
+                    local -r SPOTIFY_ACCESS_TOKEN=$( \
+                        printf "${SPOTIFY_TOKEN_RESPONSE_DATA}" | jq ".access_token" | sed 's/"//g'
```
Comment (hnarayanan) [inline spotify]: Is `jq` something that comes with every mac?

File: spotify [hunk 5]
```diff
@@ -26,6 +26,35 @@
...
 # SOFTWARE.
 
+USER_CONFIG_DEFAULTS='CLIENT_ID=""
+CLIENT_SECRET=""'
+USER_CONFIG_FILE="${HOME}/.shpotify.cfg"
+if ! [[ -f "${USER_CONFIG_FILE}" ]]; then
+    touch "${USER_CONFIG_FILE}"
+    echo "${USER_CONFIG_DEFAULTS}" > "${USER_CONFIG_FILE}"
+fi
+source "${USER_CONFIG_FILE}"
```
Comment (fonsecapeter) [inline spotify]: Just put these at the top so users can't overwrite any existing variables -- happy do something else if you have any better ideas.

Also happy to change the variable names if you want the user-interface to be any different

File: spotify [hunk 6]
```diff
@@ -26,6 +26,35 @@
...
+
+showAPIHelp() {
```
Comment (fonsecapeter) [inline spotify]: Just spit-balling here, feel free to tear this apart

File: spotify [hunk 7]
```diff
@@ -60,6 +89,7 @@ showHelp () {
     echo;
     echo "  toggle shuffle               # Toggles shuffle playback mode.";
     echo "  toggle repeat                # Toggles repeat playback mode.";
+    showAPIHelp
```
Comment (fonsecapeter) [inline spotify]: Just so people can see this on `--help` also open to removing

File: spotify [hunk 8]
```diff
@@ -119,18 +149,57 @@ while [ $# -gt 0 ]; do
...
                     cecho "Searching ${type}s for: $Q";
 
                     SPOTIFY_PLAY_URI=$( \
-                        curl -s -G $SPOTIFY_SEARCH_API --data-urlencode "q=$Q" -d "type=$type&limit=1&offset=0" -H "Accept: application/json" \
+                        curl -s -G $SPOTIFY_SEARCH_API \
+                            -H "Authorization: Bearer ${SPOTIFY_ACCESS_TOKEN}" \
+                            -H "Accept: application/json" \
+                            --data-urlencode "q=$Q" \
+                            -d "type=$type&limit=1&offset=0" \
                         | grep -E -o "spotify:$type:[a-zA-Z0-9]+" -m 1
```
Comment (fonsecapeter) [inline spotify]: I can also add a grep/error message here if the fault is in the oauth when I get some more time to work on this

File: spotify (hunk 6 above)
Comment (fonsecapeter) [reply spotify]: Also just noticed your `;`'s -- I'll go back and add those in to stay consistent when I get some more time to work on this

File: spotify (hunk 4 above)
Comment (hnarayanan) [reply spotify]: It is not a good idea because I don't use homebrew. 😅

**Comment by simoheinonen** [comment]
This is very nice!!

On macOS the default `base64` doesn't support `--wrap=0` option. The GNU one can be installed with `brew install coreutils` for example. Could consider supporting both somehow though.

```
$ base64 --help
Usage:	base64 [-hvD] [-b num] [-i in_file] [-o out_file]
  -h, --help     display this message
  -D, --decode   decodes input
  -b, --break    break encoded string into num character lines
  -i, --input    input file (default: "-" for stdin)
  -o, --output   output file (default: "-" for stdout)

$ gbase64 --help
Usage: gbase64 [OPTION]... [FILE]
Base64 encode or decode FILE, or standard input, to standard output.

With no FILE, or when FILE is -, read standard input.

Mandatory arguments to long options are mandatory for short options too.
  -d, --decode          decode data
  -i, --ignore-garbage  when decoding, ignore non-alphabet characters
  -w, --wrap=COLS       wrap encoded lines after COLS character (default 76).
                          Use 0 to disable line wrapping

      --help     display this help and exit
      --version  output version information and exit

The data are encoded as described for the base64 alphabet in RFC 4648.
When decoding, the input may contain newlines in addition to the bytes of
the formal base64 alphabet.  Use --ignore-garbage to attempt to recover
from any other non-alphabet bytes in the encoded stream.

GNU coreutils online help: <http://www.gnu.org/software/coreutils/>
Report base64 translation bugs to <http://translationproject.org/team/>
Full documentation at: <http://www.gnu.org/software/coreutils/base64>
or available locally via: info '(coreutils) base64 invocation'
```

**Comment by fonsecapeter** [comment]
Great catch @simoheinonen!

I always forget I have those coreutils, we definitely don't want to add any dependencies. I just pushed a change that uses `tr -d "\n"` instead (should cover folks who have GNU `base64` that defaults to text-wrapping). I tried both with the GNU utils and after a `brew uninstall coreutils` and everything works fine either way (at least for me)

**Comment by hnarayanan** [comment]
I am a bit confused. It looks like the discussion is taking place and you're editing your PR's code, but I am not able to see the commits here (as in it still says `Commits: 1`). How do I see the most recent code so that I can start checking it out?

**Comment by fonsecapeter** [comment]
Oh that's totally my bad, I usually amend my commit (so it just updates the one every time I push) but I'll stop doing that so it's easier to follow -- I can always squash before the merge if there end up being a bunch of unneeded commits.

Everything under "changes" (the file diff) should be up to date even though its just one commit

**Comment by hnarayanan** [comment]
Oh, ok. That is not my usual workflow so was confused.

Please, for this PR and contributing to this project, don’t squash your commits. I find it easier to track a changes commit by commit rather than one monolithic patch.

**Comment by fonsecapeter** [comment]
No worries, will do!

**Comment by fonsecapeter** [comment]
Hi @hnarayanan -- I'm sure you're super busy, but let me know if there's anything I can do to make it easier to follow my PR. I just want to help ship this so people can back to using the search feature!

I'll definitely stick to just regular commits/no squashing moving forward, but in the meantime I can add some comments to [the diff](https://github.com/hnarayanan/shpotify/pull/81/files) and annotate what's changed if that would help? 

The user-community has been super helpful in making sure everything works on vanilla osx, and that seams to be all squared away, so I think the only things outstanding are:
  1. Do you like the idea of  having a `~/.shpotify.cfg`
      - would you rather this be a pure text file that's parsed instead of sourced?
        - (sourced config=simpler to code, parsed text=maybe simpler for users)
      - would you prefer this file to "auto-touch" if it doesn't exist or let users manually touch the file if they want to use the search feature?
  2. Does the help message look good to you?
      - I tried to stay consistent with the current implementation

Anyway thanks again for making this and keeping it alive!

**Comment by hnarayanan** [comment]
Thank you!

I've merged this into a separate branch where I can play with it, clean it up if needed and then merge back into master.

**Comment by hnarayanan** [comment]
And this is now in master, and I will make a new release soon.

This was such an excellent PR that not only added a critical piece of functionality, it really thought through the UI really well and it was a pleasure to use. Thank you very much.

**Comment by fonsecapeter** [comment]
Awesome thank you! Your project is really well organized and made it easy to jump in

**Pull request merged**